cpu/sam3x8e/periph/pwm.c

```diff
@@ -0,0 +1,319 @@
+     * With every step the prescaler is doubled and the remaining quotient
+     * is halfed and tested <=255.
+     */
+    if (F_CPU < pwm_clk) {  /* Have to cut down resulting frequency. */
+        frequency = F_CPU / resolution;
+    }
+    else {   /* Estimate prescaler and divider. */
+        diva = F_CPU / pwm_clk;
+
+        while ((prea < MCK_DIV_LB_MAX) && (~0xff & diva)) {
+            prea = prea + 1;
+            diva = diva >> 1;
+        }
+
+        frequency = F_CPU / ((resolution * diva) << prea);
+    }
+
+    retval = frequency;
+
+    /* Activate PWM block by enabling it's clock. */
+    PMC->PMC_PCER1 = PMC_PCER1_PID36;
+
+    /*
+     * ++ Initialisation ++
+     * adviced by "38.6.5.1 Initialization" in
+     * "Atmel SAM3X8E  SAM3X8C SAM3X4E SAM3X4C SAM3A8C SAM3A8C Datasheet"
+     */
+    /* Unlock User Interface */
+    PWM_0_DEV->PWM_WPCR  = PWM_ENA_CHID0;
+
+    /* Disable all channels to allow CPRD updates. */
+    PWM_0_DEV->PWM_DIS = 0xff;
+
+    /* Step 2. Configure clock generator */
+    PWM_0_DEV->PWM_CLK = PWM_CLK_PREA(prea)  | PWM_CLK_DIVA(diva) |
+                         (~(PWM_CLK_PREA_Msk | PWM_CLK_DIVA_Msk)  &
+                          PWM_0_DEV->PWM_CLK);
+
+    /* +++++++++++ Configure and init channels +++++++++++++++ */
+
+    /* Set clock source, resolution, duty-cycle and enable */
+#if PWM_0_CHANNELS > 0
+    PWM_0_DEV_CH0->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    PWM_0_DEV_CH0->PWM_CPRD = resolution - 1;
+    PWM_0_DEV_CH0->PWM_CDTY = 0;
+    PWM_0_DEV->PWM_ENA = PWM_0_ENA_CH0;
+#endif
+#if PWM_0_CHANNELS > 1
+    PWM_0_DEV_CH1->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    PWM_0_DEV_CH1->PWM_CPRD = resolution - 1;
+    PWM_0_DEV_CH1->PWM_CDTY = 0;
+    PWM_0_DEV->PWM_ENA = PWM_0_ENA_CH1;
+#endif
+#if PWM_0_CHANNELS > 2
+    PWM_0_DEV_CH2->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    PWM_0_DEV_CH2->PWM_CPRD = resolution - 1;
+    PWM_0_DEV_CH2->PWM_CDTY = 0;
+    PWM_0_DEV->PWM_ENA = PWM_0_ENA_CH2;
+#endif
+#if PWM_0_CHANNELS > 3
+    PWM_0_DEV_CH3->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    PWM_0_DEV_CH3->PWM_CPRD = resolution - 1;
+    PWM_0_DEV_CH3->PWM_CDTY = 0;
+    PWM_0_DEV->PWM_ENA = PWM_0_ENA_CH3;
+#endif
+#if PWM_0_CHANNELS > 4
+    PWM_0_DEV_CH4->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    PWM_0_DEV_CH4->PWM_CPRD = resolution - 1;
+    PWM_0_DEV_CH4->PWM_CDTY = 0;
+    PWM_0_DEV->PWM_ENA = PWM_0_ENA_CH4;
+#endif
+#if PWM_0_CHANNELS > 5
+    PWM_0_DEV_CH5->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    PWM_0_DEV_CH5->PWM_CPRD = resolution - 1;
+    PWM_0_DEV_CH5->PWM_CDTY = 0;
+    PWM_0_DEV->PWM_ENA = PWM_0_ENA_CH5;
+#endif
+#if PWM_0_CHANNELS > 6
+    PWM_0_DEV_CH6->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    PWM_0_DEV_CH6->PWM_CPRD = resolution - 1;
+    PWM_0_DEV_CH6->PWM_CDTY = 0;
+    PWM_0_DEV->PWM_ENA = PWM_0_ENA_CH6;
+#endif
+#if PWM_0_CHANNELS > 7
+    PWM_0_DEV_CH7->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    PWM_0_DEV_CH7->PWM_CPRD = resolution - 1;
+    PWM_0_DEV_CH7->PWM_CDTY = 0;
+    PWM_0_DEV->PWM_ENA = PWM_0_ENA_CH7;
+#endif
```

Comment:
Looking at theses lines I don't get your macros in the periph_conf. `PWM_MAX_CHANNELS    (8U)` and `PWM_0_CHANNELS      (4U) `. In the periph_conf you introduced further macros for `PWM_0_CH_0...3`. Here you have blocks for `PWM_0_CH_0...7`. Why?